routing/cross_mwm_road_graph.cpp

```diff
@@ -157,70 +200,123 @@ bool CrossMwmGraph::ConstructBorderCrossImpl(OutgoingCrossNode const & startNode
   ASSERT(crosses.empty(), ());
   nextMapping->LoadCrossContext();
   nextMapping->m_crossContext.ForEachIngoingNodeNearPoint(startNode.m_point, [&](IngoingCrossNode const & node)
-    {
-    if (node.m_point.EqualDxDy(startNode.m_point, kMwmCrossingNodeEqualityMeters * MercatorBounds::degreeInMetres))
-    {
-      auto const toCross = CrossNode(node.m_nodeId, nextMapping->GetMwmId(), node.m_point);
-      if (toCross.IsValid())
-        crosses.emplace_back(fromCross, toCross);
-    }
+  {
+    if (node.m_nodeId == INVALID_NODE_ID)
+      return;
+
+    if (!node.m_point.EqualDxDy(startNode.m_point, kMwmCrossingNodeEqualityMeters * MercatorBounds::degreeInMetres))
+      return;
+
+    crosses.emplace_back(fromCross, CrossNode(node.m_nodeId, nextMapping->GetMwmId(), node.m_point));
   });
   return !crosses.empty();
 }
 
-void CrossMwmGraph::GetOutgoingEdgesList(BorderCross const & v,
-                                         vector<CrossWeightedEdge> & adj) const
+bool CrossMwmGraph::ConstructBorderCrossByIngoingImpl(IngoingCrossNode const & startNode,
+                                                      TRoutingMappingPtr const & currentMapping,
+                                                      vector<BorderCross> & crosses) const
+{
+  ASSERT(crosses.empty(), ());
+  auto const toCross = CrossNode(startNode.m_nodeId, currentMapping->GetMwmId(), startNode.m_point);
+  vector<string> const & neighboringMwms = currentMapping->m_crossContext.GetNeighboringMwmList();
+  string const & currentMwm = currentMapping->GetMwmId().GetInfo()->GetCountryName();
+  // Note. There's no field |m_ingoingIndex| in class IngoingCrossNode. Because this
+  // index is not saved in osrm routing section. So we need to write a work around and
+  // to check all neighboring mwms.
+  for (string const & nextMwm : neighboringMwms)
+  {
+    TRoutingMappingPtr nextMapping = m_indexManager.GetMappingByName(nextMwm);
+    if (!nextMapping->IsValid())
+      continue;
+
+    nextMapping->LoadCrossContext();
+    nextMapping->m_crossContext.ForEachOutgoingNodeNearPoint(startNode.m_point, [&](OutgoingCrossNode const & node){
+      if (node.m_nodeId == INVALID_NODE_ID)
+        return;
+
+      if (nextMapping->m_crossContext.GetOutgoingMwmName(node) != currentMwm)
```

Comment:
Which one are you talking about? ConstructBorderCrossByOutgoingImpl() method? Or ForEachOutgoingNodeNearPoint() method call?

If you mean ConstructBorderCrossByOutgoingImpl() and ConstructBorderCrossByIngoingImpl() they have a significant difference. 

In case of ConstructBorderCrossByIngoingImpl() there's data for implementation currentMapping->m_crossContext.GetOutgoingMwmName(startNode);

In case of CrossMwmGraph::ConstructBorderCrossByIngoingImpl() there's no data in osrm section for implementation something like m_crossContext.GetIngoingMwmName()